src/main/java/org/commcare/formplayer/services/MenuSessionRunnerService.java

```diff
@@ -181,6 +190,40 @@ private BaseResponseBean getNextMenu(MenuSession menuSession,
         return menuResponseBean;
     }
 
+    private BaseResponseBean getSmartResponse(MenuSession menuSession, String[] smartLinkParams) {
+        if (smartLinkParams == null || smartLinkParams.length == 0) {
+            return null;
+        }
+
+        String template = storageFactory.getPropertyManager().getSmartLinkTemplate();
+        if (template != null) {
+            SessionWrapper session = menuSession.getSessionWrapper();
+            String command = session.getCommand();
+            if (command != null) {
+                Endpoint endpoint = menuSession.getEndpointByCommand(command);
+                if (endpoint != null) {
+                    for (String param : smartLinkParams) {
+                        template = template.replaceFirst("---", param);
+                    }
+                    template = template.replace("---", endpoint.getId());
+                    HttpUrl.Builder urlBuilder = HttpUrl.parse(template).newBuilder();
+                    OrderedHashtable<String, String> data = session.getData();
```

Comment:
This doesn't work. The case id gets pushed onto the stack and then added to the session, but on the navigation request, [this block](https://github.com/dimagi/formplayer/blob/aa5ebb994318d8a920b30d4f0cb6f026a22287db/src/main/java/org/commcare/formplayer/services/MenuSessionRunnerService.java#L259-L262) executes, and ultimately it fails [this check](https://github.com/dimagi/formplayer/blob/aa5ebb994318d8a920b30d4f0cb6f026a22287db/src/main/java/org/commcare/formplayer/services/MenuSessionFactory.java#L74-L76) because the case isn't in the user's casedb.

I've been banging my head against the wall with this for a while and am thinking about just passing along any case-id-like selections as endpoint arguments, instead of pulling from the session.

Reply:
Ahh I am running into this as well for https://github.com/dimagi/formplayer/pull/1006 as the navigation to a form break after the case selection. I am going to spend some time today and try to find a solution to this.

Reply:
Talked to Clayton on this and it seems like we will need a special pathway for cross domain navigation that can move the user to the  query screen though an `action` without needing the usual `EntityScreen`.  @snopoke  Just flagging that this seems to be a bigger issue for menu navigation involving a case that user doesn't have in DB.  I am going to think more on how we can devise such an alternative route for cross domain case selection.

Reply:
@orangejenny Stumbled on an easier workaround for this which is to simply remove the `mark` and `rewind` constructs from the `stack` blocks in `action` and `remote-query` . That would mean that we always retain the intermediary query results into the session instead of cleaning them out with a `rewind`. I have not been able to think of any obvious regressions of not doing the `rewind` but do think we would need extensive case search QA to be entirely sure with this approach.

Reply:
Ahh apologies looks like this won't work out by itself as we somehow need to clear the `search_command` id from the stack for the session to move to the next step.

Reply:
Rats. I did read your first comment thinking, "Surely there must be a reason we do the rewind..."

Reply:
@orangejenny  Simon did a restructuring of app files to address this issue in https://github.com/dimagi/commcare-hq/pull/30501 , would you be able to see if you still get blocked by this issue in the new app structure for data registry.

Reply:
The smart links do now allow the user to jump to the other domain, instead of getting hung up on the navigation in the current domain. But, upon reaching the other domain, when the session is rebuilt [during endpoint navigation](https://github.com/dimagi/formplayer/blob/feac335727c77935c893325103ce19f705c65065/src/main/java/org/commcare/formplayer/services/MenuSessionRunnerService.java#L636), it gets to the point of realizing that all entries for the current datum contain a query, and so it sends the user to the case search screen. Basically, there's no mechanism in the query to say, "_Don't_ do this query if there's already a `case_id` in the session."

Reply:
I'm thinking about adding a `relevant` attribute to the `query` block, so we can skip it if the case id is already provided.

Reply:
ahh it does seem broken. What if we make `default_search`

Reply:
yes, This structure definitely makes much more sense. Overall all the changes looks great to me :)